### PR TITLE
adding asserted_encoding check on None type encoding to match text() …

### PR DESCRIPTION
**Pull request by nateprewitt**
…behavior

### CODE REVIEW

File: requests/utils.py [hunk 1]
```diff
@@ -358,13 +358,12 @@ def get_encoding_from_headers(headers):
 
 def stream_decode_response_unicode(iterator, r):
     """Stream decodes a iterator."""
+    encoding = r.encoding
 
     if r.encoding is None:
```
Comment (sigmavirus24) [inline requests/utils.py]: Why not just use `encoding` instead of `r.encoding`?

File: requests/utils.py [hunk 2]
```diff
@@ -358,13 +358,12 @@ def get_encoding_from_headers(headers):
...
-        for item in iterator:
-            yield item
-        return
+        encoding = r.apparent_encoding
 
-    decoder = codecs.getincrementaldecoder(r.encoding)(errors='replace')
+    decoder = codecs.getincrementaldecoder(encoding)(errors='replace')
```
Comment (Lukasa) [inline requests/utils.py]: Can we add the same `try...except` as in `text, except catching both`LookupError`*and*`TypeError` and, if either is raised, raising a more appropriate error? This code cannot function with a fallback, so we need to raise a useful exception here.

File: requests/utils.py (hunk 1 above)
Comment (nateprewitt) [inline requests/utils.py]: Are you referring to the reuse of "r.encoding" on line 362? I was trying to follow the pattern established with the [`text()`](https://github.com/kennethreitz/requests/blob/49b69c4751689df7aca4435ef1223a58871a64e2/requests/models.py#L770) function

File: requests/utils.py (hunk 2 above)
Comment (nateprewitt) [inline requests/utils.py]: I added in a `UnicodeError`, not sure how descriptive you'd like the text. `stream_decode_response_unicode()` doesn't seem to currently have any coverage, but I can add tests for the failures if you think that would be useful.

File: requests/utils.py (hunk 1 above)
Comment (sigmavirus24) [inline requests/utils.py]: Attribute accesses are slower than local variable look-ups though, so we might as well use `encoding` since it already exists.

Comment (nateprewitt) [inline requests/utils.py]: Alright, should be swapped to use the local `encoding` variable now.

**Comment by sigmavirus24** [comment]
@Lukasa this looks good to me. What are your thoughts?

**Comment by Lukasa** [comment]
I think this looks good to me. @sigmavirus24 are you happy to merge this directly? I think I'd call the old behaviour a bug so I'd be fine with merging this for the next release.

**Comment by sigmavirus24** [comment]
Me too. Merging!

**Pull request merged**